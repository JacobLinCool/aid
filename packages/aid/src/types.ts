export type PromiseOr<T> = T | Promise<T>;

export type LLMQuery = (
	messages: {
		role: "system" | "user" | "assistant";
		content: string;
	}[],
) => PromiseOr<string>;

/**
 * Options for an custom Aid Task.
 */
export interface AidTaskOptions<In extends string, Out> {
	/**
	 * The few-shot prompt examples.
	 */
	examples?: [string, Out][];
	/**
	 * The output schema strategy.
	 * @default "json-schema"
	 */
	strategy?: "ts" | "json-schema";
	/**
	 * Whether to check the output against the schema.
	 * @default false
	 */
	check?: boolean;
}
