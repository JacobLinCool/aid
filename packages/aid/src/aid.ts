import debug from "debug";
import type { OpenAI } from "openai";
import type { ZodIssue, z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import { printNode, zodToTs } from "zod-to-ts";
import { AidTaskOptions, LLMQuery } from "./types";

const log = debug("aid");

export class Aid {
	constructor(protected q: LLMQuery) {}

	task<In extends string, Out>(
		goal: string,
		expected: z.ZodType<Out>,
		opt?: AidTaskOptions<In, Out>,
	): (input: In) => Promise<{ result: Out; errors: ZodIssue[] }> {
		return async (input: In) => {
			const messages: { role: "system" | "user" | "assistant"; content: string }[] = [
				{
					role: "system",
					content:
						goal +
						"\nResponse in JSON format:\n" +
						(opt?.strategy === "ts"
							? printNode(zodToTs(expected).node)
							: JSON.stringify(zodToJsonSchema(expected), null, 2)),
				},
			];

			if (opt?.examples) {
				for (const [input, output] of opt.examples) {
					messages.push({ role: "user", content: input });
					messages.push({ role: "system", content: JSON.stringify(output) });
				}
			}
			messages.push({ role: "user", content: input });

			log("messages", messages);
			const output = await this.q(messages);
			log("output", output);
			const json = JSON.parse(output);

			if (!opt?.check) {
				return { result: json, errors: [] };
			}

			const result = expected.safeParse(json);
			if (result.success) {
				return { result: json, errors: [] };
			} else {
				return { result: json, errors: result.error.issues };
			}
		};
	}

	static from(
		openai: OpenAI,
		param: Omit<OpenAI.Chat.ChatCompletionCreateParamsNonStreaming, "messages">,
	): Aid {
		const q: LLMQuery = async (
			messages: { role: "system" | "user" | "assistant"; content: string }[],
		) => {
			const payload: OpenAI.Chat.ChatCompletionCreateParamsNonStreaming = {
				temperature: 0,
				...param,
				messages,
				response_format: { type: "json_object" },
			};

			const res = await openai.chat.completions.create(payload);
			log({ fingerprint: res.system_fingerprint, usage: res.usage });

			const text = res.choices[0].message.content;
			if (!text) {
				throw new Error("No text returned from OpenAI");
			}

			return text;
		};
		return new Aid(q);
	}
}
