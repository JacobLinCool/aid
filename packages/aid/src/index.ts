export * from "./aid";
export * from "./types";
