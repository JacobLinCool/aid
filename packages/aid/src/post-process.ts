// post process the output from the large language model, and make it parseable by the JSON parser
export function post_process(llm_output: string): string {
	llm_output = llm_output.trim();

	// remove possible markdown code fence at the beginning and end
	llm_output = llm_output.replace(/^```.*\n/m, "").replace(/```\n?$/m, "");

	return llm_output;
}
