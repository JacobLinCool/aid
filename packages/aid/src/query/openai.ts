import debug from "debug";
import type { OpenAI } from "openai";
import type { LLMQuery, Message } from "../types";

const log = debug("aid:openai");

export const OpenAIQuery = (
	openai: OpenAI,
	param: Omit<OpenAI.Chat.ChatCompletionCreateParamsNonStreaming, "messages">,
): LLMQuery => {
	const q: LLMQuery = async (messages: Message[]) => {
		const payload: OpenAI.Chat.ChatCompletionCreateParamsNonStreaming = {
			temperature: 0,
			...param,
			messages,
			response_format: { type: "json_object" },
		};

		log("payload", payload);
		const res = await openai.chat.completions.create(payload);
		log({ id: res.id, fingerprint: res.system_fingerprint, usage: res.usage });

		const text = res.choices[0].message.content;
		if (!text) {
			throw new Error("No text returned from OpenAI");
		}
		log("text", text);

		return text;
	};
	return q;
};
