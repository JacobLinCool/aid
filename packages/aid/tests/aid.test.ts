import debug from "debug";
import { OpenAI } from "openai";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { Aid } from "../src";
import { ESSAY, WIKI_AI } from "./fixtures/articles";

debug.enable("aid");

describe("Aid", () => {
	it(
		"Type Constraints",
		async (c) => {
			const api_key = process.env.OPENAI_API_KEY;
			if (!api_key) {
				c.skip();
				return;
			}

			const openai = new OpenAI({ apiKey: api_key });
			const aid = Aid.from(openai, {
				model: "gpt-4-1106-preview",
				seed: 20030317,
			});

			const analyze = aid.task(
				"Summarize and extract keywords",
				z.object({
					summary: z.string().max(300),
					keywords: z.array(z.string().max(30)).max(10),
				}),
			);

			const { result } = await analyze(WIKI_AI);
			expect(typeof result).toBe("object");
			expect(typeof result.summary).toBe("string");
			expect(Array.isArray(result.keywords)).toBe(true);
			result.keywords.forEach((keyword) => expect(typeof keyword).toBe("string"));
			// expect(result).toMatchSnapshot();
		},
		{ timeout: 60_000 },
	);

	it(
		"Few-Shot Learning",
		async (c) => {
			const api_key = process.env.OPENAI_API_KEY;
			if (!api_key) {
				c.skip();
				return;
			}

			const openai = new OpenAI({ apiKey: api_key });
			const aid = Aid.from(openai, {
				model: "gpt-4-1106-preview",
				seed: 20030317,
			});

			// question source: https://www.gept.org.tw/2022/geptscoreremark/icomposition.pdf
			const grade = aid.task(
				`主題：一般來說，孩子表現良好時，父母常會給孩子獎勵。請寫一篇文章說明 (1) 你表現好的時候，你的父母通常會用哪些方法獎勵你？你覺得這些方法有效、適當嗎？ (2) 如果有一天你為人父母，你會用相同的獎勵方式嗎？` +
					`評分重點包括內容、組織、文法、用字遣詞、標點 符號、大小寫。`,
				z.object({
					grade: z.number().int().min(1).max(5),
					feedback: z.string().max(300),
				}),
				{
					examples: [
						[ESSAY[5][0], { grade: 5, feedback: ESSAY[5][1] }],
						[ESSAY[4][0], { grade: 4, feedback: ESSAY[4][1] }],
						[ESSAY[2][0], { grade: 2, feedback: ESSAY[2][1] }],
						[ESSAY[3][0], { grade: 3, feedback: ESSAY[3][1] }],
					],
				},
			);

			const { result: result1 } = await grade(
				`When I exhibited good behavior, my parents typically rewarded me with verbal praise and extra privileges, like additional playtime or a special meal. These methods were effective, making me feel appreciated and teaching the value of hard work. As a future parent, I plan to use similar strategies but with a modern twist. I might involve children in setting their own goals and rewards, fostering autonomy and understanding of their actions. Adapting these techniques to suit individual needs and changing times seems essential.`,
			);
			expect(typeof result1).toBe("object");
			expect(typeof result1.grade).toBe("number");
			expect(typeof result1.feedback).toBe("string");
			// expect(result1).toMatchSnapshot();

			const { result: result2 } = await grade(
				`When I was good, my parents gave rewards like saying 'good job' or letting me play more. I think it worked okay. If I'm a parent, I'll probably do the same but maybe let kids choose their rewards. It's important to change with the times.`,
			);
			expect(typeof result2).toBe("object");
			expect(typeof result2.grade).toBe("number");
			expect(typeof result2.feedback).toBe("string");
			// expect(result2).toMatchSnapshot();
		},
		{ timeout: 60_000 },
	);
});
