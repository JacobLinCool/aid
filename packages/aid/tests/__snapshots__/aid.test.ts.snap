// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aid > Few-Shot Learning 1`] = `
{
  "feedback": "文章內容完整，清楚回答了題目的兩個部分。組織結構良好，先描述自己的經歷，然後談論未來的計劃。文法正確，用字遣詞恰當，例如使用了\\"exhibited good behavior\\"和\\"fostering autonomy\\"等精確詞彙。標點符號和大小寫使用正確，整體表達流暢，沒有明顯錯誤。",
  "grade": 5,
}
`;

exports[`Aid > Few-Shot Learning 2`] = `
{
  "feedback": "內容簡潔，但基本符合題目要求。組織合理，但篇幅較短，未能深入探討獎勵方式的有效性或適當性。文法和用字遣詞基本正確，但表達方式過於簡單，缺乏豐富的詞彙和複雜句型。建議未來作文時能夠提供更多細節和例子，以增強文章的說服力。",
  "grade": 3,
}
`;

exports[`Aid > Type Constraints 1`] = `
{
  "keywords": [
    "artificial intelligence",
    "AI",
    "machine intelligence",
    "computer science",
    "deep learning",
    "search engines",
    "recommendation systems",
    "voice assistants",
    "self-driving cars",
    "natural language processing",
  ],
  "summary": "Artificial intelligence (AI) is the simulation of human intelligence in machines, aimed at performing tasks that typically require human intelligence. It encompasses a variety of applications such as search engines, recommendation systems, voice assistants, self-driving cars, and gaming. AI research, which began in 1956, has experienced ups and downs but saw a resurgence with the advent of deep learning in 2012. AI research includes reasoning, learning, perception, and more, utilizing tools from various disciplines including computer science, psychology, and neuroscience.",
}
`;
